examples/all-clusters-app/all-clusters-common/shell_extension/control.cpp

```diff
@@ -0,0 +1,260 @@
+#include "app/util/DataModelHandler.h"
+#include "app/util/basic-types.h"
+#include "core/CHIPError.h"
+#include "core/NodeId.h"
+#include "messaging/ExchangeContext.h"
+#include "messaging/ExchangeDelegate.h"
+#include "protocols/secure_channel/SessionEstablishmentDelegate.h"
+#include "support/logging/CHIPLogging.h"
+#include "transport/FabricTable.h"
+#include "transport/SessionHandle.h"
+#include "transport/raw/PeerAddress.h"
+#include <app/server/Server.h>
+#include <bits/stdint-uintn.h>
+#include <controller/CHIPDevice.h>
+#include <controller/InteractionPeer.h>
+#include <credentials/CHIPCert.h>
+#include <cstdio>
+#include <lib/shell/commands/Help.h>
+#include <lib/shell/streamer.h>
+#include <lib/support/BytesToHex.h>
+#include <lib/support/CodeUtils.h>
+#include <memory>
+#include <protocols/secure_channel/CASESession.h>
+#include <zap-generated/CHIPClusters.h>
+
+using chip::CASESession;
+using chip::FabricIndex;
+using chip::NodeId;
+using chip::Optional;
+using chip::SecureSessionMgr;
+using chip::Server;
+using chip::SessionEstablishmentDelegate;
+using chip::SessionHandle;
+using chip::Controller::InteractionPeer;
+using chip::Controller::InteractionPeerDelegate;
+using chip::Messaging::ExchangeContext;
+using chip::Messaging::ExchangeDelegate;
+using chip::Shell::Engine;
+using chip::Shell::PrintCommandHelp;
+using chip::Shell::shell_command_t;
+using chip::Shell::streamer_get;
+using chip::Shell::streamer_printf;
+using chip::Transport::FabricInfo;
+using chip::Transport::PeerAddress;
+
+namespace {
+
+class Device : public InteractionPeerDelegate, public SessionEstablishmentDelegate, public ExchangeDelegate
+{
+public:
+    Device(NodeId nodeId, PeerAddress address, FabricIndex fabricIndex) :
+        mPeerId(nodeId), mPeerAddress(address), mFabricIndex(fabricIndex), mInteractionPeer(this)
+    {}
+
+    CHIP_ERROR Connect()
+    {
+        uint16_t keyID = 0;
+        ReturnErrorOnFailure(Server::GetServer().GetSessionIDAllocator().Allocate(keyID));
+        ExchangeContext * exchange = Server::GetServer().GetExchangManager().NewContext(SessionHandle(), &mCASESession);
+        VerifyOrReturnError(exchange != nullptr, CHIP_ERROR_INTERNAL);
+        ReturnErrorOnFailure(
+            mCASESession.MessageDispatch().Init(Server::GetServer().GetSecureSessionManager().GetTransportManager()));
+        mCASESession.MessageDispatch().SetPeerAddress(mPeerAddress);
+        FabricInfo * fabric = Server::GetServer().GetFabricTable().FindFabricWithIndex(mFabricIndex);
+        ReturnErrorCodeIf(fabric == nullptr, CHIP_ERROR_INCORRECT_STATE);
+        ReturnErrorOnFailure(mCASESession.EstablishSession(mPeerAddress, fabric, mPeerId, keyID, exchange, this));
+        return CHIP_NO_ERROR;
+    };
+
+    CHIP_ERROR ToggleOnOff(void)
+    {
+        mOnOffCluster.Associate(&mInteractionPeer, 1);
+        return mOnOffCluster.Toggle(nullptr, nullptr);
+    }
+
+private:
+    CHIP_ERROR PreparePeer() override { return CHIP_NO_ERROR; };
```

Comment:
The logic of `PreparePeer` is complicated in `Controller::Device` since it needs to handle PASESessions and the state change. As we are directly setting up the CASESession here this function can be left empty. Ideally the session type should not affect the behavior of data model clients. However this will also require us to rewrite the `Device` and the `DeviceController` class.

Reply:
Why not just move the contents of `Connect()` to `PreparePeer()`? Is there something about the intention behind `PreparePeer()` that means it should not do CASE setup?